cmd/ceremony/main.go

```diff
@@ -22,6 +23,156 @@ import (
 	"gopkg.in/yaml.v3"
 )
 
+// issueLintCert issues a linting certificate from a given template certificate
+// signed by a given issuer and returns the certificate or an error. The public
+// key from the loaded certificate is checked by the GoodKey package.
+func issueLintCert(tbs, issuer *x509.Certificate, subjectPubKey crypto.PublicKey, signer crypto.Signer, skipLints []string) (*x509.Certificate, error) {
+	lintCertBytes, err := linter.Check(tbs, subjectPubKey, issuer, signer, skipLints)
+	if err != nil {
+		return nil, fmt.Errorf("certificate failed pre-issuance lint: %w", err)
+	}
+
+	lintCert, err := x509.ParseCertificate(lintCertBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	err = kp.GoodKey(context.Background(), lintCert.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return lintCert, nil
+}
+
+func equalPubKeys(a, b interface{}) bool {
+	aBytes, err := x509.MarshalPKIXPublicKey(a)
+	if err != nil {
+		return false
+	}
+	bBytes, err := x509.MarshalPKIXPublicKey(b)
+	if err != nil {
+		return false
+	}
+	return bytes.Equal(aBytes, bBytes)
+}
+
+func openSigner(cfg PKCS11SigningConfig, pubKey crypto.PublicKey) (crypto.Signer, *hsmRandReader, error) {
+	session, err := pkcs11helpers.Initialize(cfg.Module, cfg.SigningSlot, cfg.PIN)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to setup session and PKCS#11 context for slot %d: %s",
+			cfg.SigningSlot, err)
+	}
+	log.Printf("Opened PKCS#11 session for slot %d\n", cfg.SigningSlot)
+	signer, err := session.NewSigner(cfg.SigningLabel, pubKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to retrieve private key handle: %s", err)
+	}
+	if !equalPubKeys(signer.Public(), pubKey) {
+		return nil, nil, fmt.Errorf("signer pubkey did not match issuer pubkey")
+	}
+	log.Println("Retrieved private key handle")
+	return signer, newRandReader(session), nil
+}
+
+// loadCert loads a PEM certificate specified by filename or returns an error.
+// The public key from the loaded certificate is checked by the GoodKey package.
+func loadCert(filename string) (*x509.Certificate, error) {
+	certPEM, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	log.Printf("Loaded certificate from %s\n", filename)
+	block, _ := pem.Decode(certPEM)
+	if block == nil {
+		return nil, fmt.Errorf("No data in cert PEM file %s", filename)
+	}
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return nil, err
+	}
+	goodkeyErr := kp.GoodKey(context.Background(), cert.PublicKey)
+	if goodkeyErr != nil {
+		return nil, goodkeyErr
+	}
+
```

Comment:
nit: either consistently have a blank line after every `if err != nil` block, or don't have space after any of them.

Reply:
The particular blank line follows the same style pattern elsewhere in boulder - a blank line before the final `return` of a function/method. I'll see where else I have a mix of blank lines/condensed code and make it all cohesive.